File: C++/0121-best-time-stock.cpp
//time O(n)
//space O(1)
class Solution {
public:
    int maxProfit(vector<int>& prices) {


      if(prices.empty())
      return 0;

      int minPrice = numeric_limits<int>::max();
      int maxProfit = 0;

      for(const auto &p : prices){
        if(p < minPrice)
        minPrice = p;
        else if (p-minPrice > maxProfit)
        maxProfit = p - minPrice;
      }
        return maxProfit;
    }
};

File: Python/0121-best-time-stock.py
# Time O(n)
# Space O(1)
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        maxProfit, minPrice = 0, float("inf")

        for p in prices:
            minPrice = min(minPrice,p)
            maxProfit = max(maxProfit,p-minPrice)

        return maxProfit
